libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+    where
+        I: IntoIterator<Item = &'a Topic>,
+        I::IntoIter: Clone,
+    {
+        // This function exists for convenience.
+        self.sub_unsub_multi(topics.into_iter().map::<_, fn(_) -> _>(|t| (t, false)));
+    }
+
+    // Inner implementation. The iterator should produce a boolean that is true if we subscribe and
+    // false if we unsubscribe.
+    fn sub_unsub_multi<'a, I>(&self, topics: I)
+    where
+        I: IntoIterator<Item = (&'a Topic, bool)>,
+        I::IntoIter: Clone,
+    {
+        let mut proto = rpc_proto::RPC::new();
+
+        let topics = topics.into_iter();
+
+        debug!(target: "libp2p-floodsub", "Queuing sub/unsub message ; sub = {:?} ; unsub = {:?}",
+               topics.clone().filter(|t| t.1)
+                     .map(|t| t.0.hash().clone().into_string())
+                     .collect::<Vec<_>>(),
+               topics.clone().filter(|t| !t.1)
+                     .map(|t| t.0.hash().clone().into_string())
+                     .collect::<Vec<_>>());
+
+        let mut subscribed_topics = self.inner.subscribed_topics.write();
+        for (topic, subscribe) in topics {
+            let mut subscription = rpc_proto::RPC_SubOpts::new();
+            subscription.set_subscribe(subscribe);
+            subscription.set_topicid(topic.hash().clone().into_string());
+            proto.mut_subscriptions().push(subscription);
+
+            if subscribe {
+                subscribed_topics.push(topic.clone());
+            } else {
+                subscribed_topics.retain(|t| t.hash() != topic.hash())
+            }
+        }
+
+        self.broadcast(proto, |_| true);
+    }
+
+    /// Publishes a message on the network for the specified topic
+    #[inline]
+    pub fn publish(&self, topic: &Topic, data: Vec<u8>) {
+        // This function exists for convenience.
+        self.publish_multi(iter::once(topic), data)
+    }
+
+    /// Publishes a message on the network for the specified topics.
+    pub fn publish_multi<'a, I>(&self, topics: I, data: Vec<u8>)
+    where
+        I: IntoIterator<Item = &'a Topic>,
+        I::IntoIter: Clone,
+    {
+        let topics = topics.into_iter();
+
+        debug!(target: "libp2p-floodsub", "Queueing publish message ; \
+                                           topics = {:?} ; data_len = {:?}",
+               topics.clone().map(|t| t.hash().clone().into_string()).collect::<Vec<_>>(),
+               data.len());
+
+        // Build the `Vec<u8>` containing our sequence number for this message.
+        let seq_no_bytes = {
+            let mut seqno_bytes = Vec::new();
+            let seqn = self.inner.seq_no.fetch_add(1, Ordering::Relaxed);
+            seqno_bytes
+                .write_u64::<BigEndian>(seqn as u64)
+                .expect("writing to a Vec never fails");
+            seqno_bytes
+        };
+
+        // TODO: should handle encryption/authentication of the message
+
+        let mut msg = rpc_proto::Message::new();
```

Comment:
Using `derive(Serialize, Deserialize)` and another format would also make this code _way_ less ugly.